networking/eth2/src/main/java/tech/pegasys/artemis/networking/eth2/gossip/topics/validation/SignedAggregateAndProofValidator.java

```diff
@@ -0,0 +1,176 @@
+import static tech.pegasys.artemis.util.config.Constants.DOMAIN_SELECTION_PROOF;
+import static tech.pegasys.artemis.util.config.Constants.VALID_BLOCK_SET_SIZE;
+
+import com.google.common.primitives.UnsignedLong;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.tuweni.bytes.Bytes;
+import tech.pegasys.artemis.bls.BLS;
+import tech.pegasys.artemis.bls.BLSPublicKey;
+import tech.pegasys.artemis.bls.BLSSignature;
+import tech.pegasys.artemis.datastructures.operations.AggregateAndProof;
+import tech.pegasys.artemis.datastructures.operations.Attestation;
+import tech.pegasys.artemis.datastructures.operations.SignedAggregateAndProof;
+import tech.pegasys.artemis.datastructures.state.BeaconState;
+import tech.pegasys.artemis.datastructures.util.CommitteeUtil;
+import tech.pegasys.artemis.datastructures.util.ValidatorsUtil;
+import tech.pegasys.artemis.storage.client.RecentChainData;
+import tech.pegasys.artemis.util.collections.ConcurrentLimitedSet;
+import tech.pegasys.artemis.util.collections.LimitStrategy;
+import tech.pegasys.artemis.util.config.Constants;
+
+public class SignedAggregateAndProofValidator {
+  private static final Logger LOG = LogManager.getLogger();
+  private final Set<AggregatorIndexAndSlot> receivedValidAggregations =
+      ConcurrentLimitedSet.create(VALID_BLOCK_SET_SIZE, LimitStrategy.DROP_OLDEST_ELEMENT);
+  private final AttestationValidator attestationValidator;
+  private final RecentChainData recentChainData;
+
+  public SignedAggregateAndProofValidator(
+      final AttestationValidator attestationValidator, final RecentChainData recentChainData) {
+    this.attestationValidator = attestationValidator;
+    this.recentChainData = recentChainData;
+  }
+
+  public ValidationResult validate(final SignedAggregateAndProof signedAggregate) {
+    final AggregateAndProof aggregateAndProof = signedAggregate.getMessage();
+    final Attestation aggregate = aggregateAndProof.getAggregate();
+
+    final UnsignedLong aggregateSlot = aggregate.getData().getSlot();
+    final AggregatorIndexAndSlot aggregatorIndexAndSlot =
+        new AggregatorIndexAndSlot(aggregateAndProof.getIndex(), aggregateSlot);
+    if (receivedValidAggregations.contains(aggregatorIndexAndSlot)) {
+      LOG.trace("Rejecting duplicate aggregate");
+      return INVALID;
+    }
+
+    final ValidationResult aggregateValidationResult =
+        attestationValidator.singleOrAggregateAttestationChecks(aggregate);
+    if (aggregateValidationResult == INVALID) {
+      LOG.trace("Rejecting aggregate because attestation failed validation");
+      return aggregateValidationResult;
+    }
+
+    // State must be present because the aggregate is valid
+    final BeaconState state =
+        recentChainData.getBlockState(aggregate.getData().getBeacon_block_root()).orElseThrow();
```

Comment:
Shouldn't we return `SAVED_FOR_FUTURE` here if the state isn't available?

Reply:
Ooo, good spot. That used to be safe because we returned the attestationValidator result but we now see if we can find some reason to upgrade from SAVED to INVALID.